lntest/itest/lnd_forward_interceptor_test.go

```diff
@@ -0,0 +1,389 @@
+	}
+
+	// prepare the test cases.
+	testCases, err := testContext.prepareTestCases()
+	if err != nil {
+		t.Fatalf("failed to prepare test cases")
+	}
+
+	// a channel for the interceptor go routine to send the requested packets.
+	interceptedChan := make(chan *routerrpc.ForwardHtlcInterceptRequest,
+		len(testCases))
+
+	// run the interceptor loop in its own go routine.
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			request, err := interceptor.Recv()
+			if err != nil {
+				return
+			}
+			interceptedChan <- request
+		}
+	}()
+
+	// For each test case make sure we initiate a payment from Alice to Carol
+	// routed through Bob. For each payment we also test its final status
+	// according to the interceptorAction specified in the test.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for _, tc := range testCases {
+			attempt, err := testContext.sendAliceToCarolPayment(
+				context.Background(), tc.invoice.ValueMsat, tc.invoice.RHash)
+
+			if t.t.Failed() {
+				return
+			}
+			if err != nil {
+				t.Fatalf("failed to send payment %v", err)
+			}
+
+			switch tc.interceptorAction {
+			// for 'fail' interceptor action we make sure the payment failed.
+			case routerrpc.ResolveHoldForwardAction_FAIL:
+				if attempt.Status != lnrpc.HTLCAttempt_FAILED {
+					t.Fatalf("expected payment to fail, instead got %v", attempt.Status)
+				}
+			// for settle and resume we make sure the payment is successfull.
+			case routerrpc.ResolveHoldForwardAction_SETTLE:
+				fallthrough
+			case routerrpc.ResolveHoldForwardAction_RESUME:
+				if attempt.Status != lnrpc.HTLCAttempt_SUCCEEDED {
+					t.Fatalf("expected payment to succeed, instead got %v", attempt.Status)
+				}
+			}
+		}
+	}()
+
+	// We make sure here the interceptor has processed all packets before we
+	// check the payment statuses.
+	for i := 0; i < len(testCases); i++ {
+		select {
+		case request := <-interceptedChan:
+			testCase := testCases[i]
+
+			// for held packets we ignore, keeping them in hold status.
+			if testCase.shouldHold {
+				continue
+			}
+
+			// for all other packets we resolve according to the test case.
+			switch testCase.interceptorAction {
+			case routerrpc.ResolveHoldForwardAction_FAIL:
+				interceptor.Send(&routerrpc.ForwardHtlcInterceptResponse{
+					IncomingCircuitKey: request.IncomingCircuitKey,
+					Action:             routerrpc.ResolveHoldForwardAction_FAIL,
+				})
+			case routerrpc.ResolveHoldForwardAction_RESUME:
+				interceptor.Send(&routerrpc.ForwardHtlcInterceptResponse{
+					IncomingCircuitKey: request.IncomingCircuitKey,
+					Action:             routerrpc.ResolveHoldForwardAction_RESUME,
+				})
+			case routerrpc.ResolveHoldForwardAction_SETTLE:
+				interceptor.Send(&routerrpc.ForwardHtlcInterceptResponse{
+					IncomingCircuitKey: request.IncomingCircuitKey,
+					Action:             routerrpc.ResolveHoldForwardAction_SETTLE,
+					Preimage:           testCase.invoice.RPreimage,
+				})
+			}
```

Comment:
Is this switch needed at all? Maybe the test case parameters can be copied straight into the `ForwardHtlcInterceptResponse`?

Reply:
Yes, much cleaner. Done.